### PR TITLE
Add support for Shelly Duo RGBW light bulbs

### PR DESCRIPTION
**Pull request by Xinayder**
This PR adds support for Shelly Duo RGBW light bulbs. Currently only tested with a GU10.

`KEY_EFFECT_TEMPLATE` was added for Lovelace to properly display what effect is currently selected. I used the Shelly RGBW2 as a base, but had to change `KEY_BRIGHTNESS_VALUE_TEMPLATE` to `KEY_BRIGHTNESS_TEMPLATE` for the same reason previously described. The light bulb only supports 4 effects, including "Off" (no effect).

To turn on the light bulb, the `effect` field must be present on the MQTT payload, thus if you only want to turn on the light and not use any effects, just set the field to 0.

Thanks to @tc-maxx for providing a MQTT discovery payload for this light bulb (#424)

This is the YAML generated by the Python script:

```yaml
schema: template
qos: 0
red_template: '{%if value_json.mode == "color"%}{{value_json.red}}{%else%}{{ none }}{%endif%}'
green_template: >-
  {%if value_json.mode == "color"%}{{value_json.green}}{%else%}{{ none
  }}{%endif%}
blue_template: >-
  {%if value_json.mode == "color"%}{{value_json.blue}}{%else%}{{ none
  }}{%endif%}
state_topic: shellies/shellycolorbulb-485519DA73A9/color/0/status
brightness_state_topic: shellies/shellycolorbulb-485519DA73A9/color/0/status
origin:
  name: Shellies Discovery
  sw_version: 5.1.0
  support_url: https://github.com/bieniu/ha-shellies-discovery
effect_command_template: >-
  { {%if value=="Off"%}"effect":0{%elif value=="Meteor Shower"%}"effect":1{%elif
  value=="Gradual Change"%}"effect":2{%elif value=="Flash"%}"effect":3{%endif%}
  }
state_template: '{%if value_json.ison%}on{%else%}off{%endif%}'
unique_id: shellycolorbulb-485519da73a9-light-0
effect_state_topic: shellies/shellycolorbulb-485519DA73A9/color/0/status
availability:
  - payload_not_available: 'false'
    topic: shellies/shellycolorbulb-485519DA73A9/online
    payload_available: 'true'
  - value_template: '{%if value_json.mqtt.connected%}online{%else%}offline{%endif%}'
    topic: shellies/shellycolorbulb-485519DA73A9/info
max_mireds: 333
min_mireds: 154
command_on_template: >-
  {"turn":"on"{%if red is defined and green is defined and blue is
  defined%},"mode":"color","red":{{red}},"green":{{green}},"blue":{{blue}}{%endif%}{%if
  brightness is
  defined%},"brightness":{{brightness|float|multiply(0.3922)|round}},"gain":{{brightness|float|multiply(0.3922)|round}}{%endif%}{%if
  color_temp is
  defined%},"mode":"white","temp":{{(1000000/(color_temp|int))|round(0,"floor")}}{%endif%}{%if
  transition is defined%},"transition":{{min(transition|multiply(1000),
  5000)}}{%endif%}{%if effect is defined%},"effect": {%- if effect == "Meteor
  Shower" -%}1{%- elif effect == "Gradual Change" -%}2{%- elif effect == "Flash"
  -%}3{%- else -%}0{%- endif -%}{%- else -%},"effect": 0{%endif%} }
brightness_command_topic: shellies/shellycolorbulb-485519DA73A9/color/0/set
effect_list:
  - 'Off'
  - Meteor Shower
  - Gradual Change
  - Flash
brightness_template: '{{value_json.brightness|float|multiply(2.55)|round(0)}}'
effect_command_topic: shellies/shellycolorbulb-485519DA73A9/color/0/set
brightness_command_template: >-
  {"gain":{{value|float|multiply(0.3922)|round(0)}},
  "brightness":{{value|float|multiply(0.3922)|round(0)}}}
device:
  name: Shelly Duo RGBW 485519DA73A9
  connections:
    - - mac
      - 48:55:19:da:73:a9
  hw_version: gen1 (SHCB-1)
  sw_version: 20231107-162425/v1.14.1-rc1-g0617c15
  model: Shelly Duo RGBW
  configuration_url: http://192.168.1.138/
  manufacturer: Allterco Robotics
command_off_template: >-
  {"turn":"off","effect":0{%if transition is
  defined%},"transition":{{min(transition|multiply(1000),5000)}}{%endif%}}
effect_template: >-
  {%if value_json.effect==1%}Meteor Shower{%elif value_json.effect==2%}Gradual
  Change{%elif value_json.effect==3%}Flash{%else%}Off{%endif%}
name: Light 0
command_topic: shellies/shellycolorbulb-485519DA73A9/color/0/set
color_temp_template: '{{((1000000/(value_json.temp|int,2700)|max)|round(0,"floor"))}}'
platform: mqtt
```

image image

Closes https://github.com/bieniu/ha-shellies-discovery/issues/588

### CODE REVIEW

File: python_scripts/shellies_discovery.py [hunk 1]
```diff
@@ -3254,6 +3278,39 @@ def mqtt_publish(topic, payload, retain, json=False):
...
+            KEY_COMMAND_OFF_TEMPLATE: f"{{^turn^:^off^,^effect^:0{{%if transition is defined%}},^transition^:{{{{min(transition|multiply(1000),{MAX_TRANSITION})}}}}{{%endif%}}}}",
+            KEY_COMMAND_TOPIC: set_topic,
+            KEY_MAX_MIREDS: 333,
+            KEY_MIN_MIREDS: 154,
+            KEY_STATE_TOPIC: status_topic,
+            "red_template": "{%if value_json.mode == ^color^%}{{value_json.red}}{%else%}{{ none }}{%endif%}",
```
Comment (bieniu) [inline python_scripts/shellies_discovery.py:3309]: Please introduce `KEY_RED_TEMPLATE` constant with an abbreviation `r_tpl`.

File: python_scripts/shellies_discovery.py [hunk 2]
```diff
@@ -3254,6 +3278,39 @@ def mqtt_publish(topic, payload, retain, json=False):
...
+            "green_template": "{%if value_json.mode == ^color^%}{{value_json.green}}{%else%}{{ none }}{%endif%}",
```
Comment (bieniu) [inline python_scripts/shellies_discovery.py:3310]: Please introduce `KEY_GREEN_TEMPLATE` constant with an abbreviation `g_tpl`.

File: python_scripts/shellies_discovery.py [hunk 3]
```diff
@@ -3254,6 +3278,39 @@ def mqtt_publish(topic, payload, retain, json=False):
...
+            "blue_template": "{%if value_json.mode == ^color^%}{{value_json.blue}}{%else%}{{ none }}{%endif%}",
```
Comment (bieniu) [inline python_scripts/shellies_discovery.py:3311]: Please introduce `KEY_BLUE_TEMPLATE` constant with an abbreviation `b_tpl`.

File: python_scripts/shellies_discovery.py [hunk 4]
```diff
@@ -3254,6 +3278,39 @@ def mqtt_publish(topic, payload, retain, json=False):
...
+            KEY_EFFECT_TEMPLATE: "{%if value_json.effect==1%}Meteor Shower{%elif value_json.effect==2%}Gradual Change{%elif value_json.effect==3%}Flash{%else%}Off{%endif%}",
+            KEY_UNIQUE_ID: unique_id,
+            KEY_QOS: qos,
+            KEY_DEVICE: device_info,
+            KEY_ORIGIN: origin_info,
+            KEY_COLOR_TEMP_TEMPLATE: TPL_COLOR_TEMP_WHITE_LIGHT,
+            KEY_COMMAND_ON_TEMPLATE: f"{{^turn^:^on^{{%if red is defined and green is defined and blue is defined%}},^mode^:^color^,^red^:{{{{red}}}},^green^:{{{{green}}}},^blue^:{{{{blue}}}}{{%endif%}}{{%if brightness is defined%}},^brightness^:{{{{brightness|float|multiply(0.3922)|round}}}},^gain^:{{{{brightness|float|multiply(0.3922)|round}}}}{{%endif%}}{{%if color_temp is defined%}},^mode^:^white^,^temp^:{{{{(1000000/(color_temp|int))|round(0,^floor^)}}}}{{%endif%}}{{%if transition is defined%}},^transition^:{{{{min(transition|multiply(1000), {MAX_TRANSITION})}}}}{{%endif%}}{{%if effect is defined%}},^effect^: {{%- if effect == ^Meteor Shower^ -%}}1{{%- elif effect == ^Gradual Change^ -%}}2{{%- elif effect == ^Flash^ -%}}3{{%- else -%}}0{{%- endif -%}}{{%- else -%}},^effect^: 0{{%endif%}} }}",
```
Comment (bieniu) [inline python_scripts/shellies_discovery.py:3303]: Please remove unneeded spaces and `-`.

File: python_scripts/shellies_discovery.py [hunk 5]
```diff
@@ -3254,6 +3278,39 @@ def mqtt_publish(topic, payload, retain, json=False):
...
+            KEY_SCHEMA: VALUE_TEMPLATE,
+            KEY_NAME: light_name,
+            KEY_AVAILABILITY: availability,
+            KEY_COMMAND_TOPIC: set_topic,
+            KEY_STATE_TOPIC: state_topic,
+            KEY_STATE_TEMPLATE: "{%if value_json.ison%}on{%else%}off{%endif%}",
+            KEY_BRIGHTNESS_STATE_TOPIC: status_topic,
+            KEY_BRIGHTNESS_TEMPLATE: "{{value_json.gain|float|multiply(2.55)|round(0)}}",
+            KEY_BRIGHTNESS_COMMAND_TOPIC: set_topic,
+            KEY_BRIGHTNESS_COMMAND_TEMPLATE: "{^gain^:{{value|float|multiply(0.3922)|round(0)}}, ^brightness^:{{value|float|multiply(0.3922)|round(0)}}}",
```
Comment (bieniu) [inline python_scripts/shellies_discovery.py:3292]:
```suggestion
            KEY_BRIGHTNESS_COMMAND_TEMPLATE: "{^gain^:{{value|float|multiply(0.3922)|round(0)}},^brightness^:{{value|float|multiply(0.3922)|round(0)}}}",
```

**Review by bieniu — CHANGES_REQUESTED** [review]
Please fix lint errors. My RGBW2 died so I need a while to get a new one to test this. Thanks!

**Comment by tcsmaxx** [comment]
Hi @Xinayder, it looks like there are still some requested changes for this PR. Would you be able to take a look and help implement the necessary adjustments? Let me know if there's anything I can assist with to move this forward. Thanks for your work on this!

**Review by bieniu — APPROVED** [review]

**Comment by Xinayder** [comment]
My notifications were off for some reason, I am terribly sorry for forgetting about this PR. I'm happy that it was merged!

**Comment by bieniu** [comment]
Thanks for your contribution 👍

**Pull request merged**